### PR TITLE
SPARK-22896 Improvement in String interpolation

### PR DESCRIPTION
**Pull request by chetkhatri**
## What changes were proposed in this pull request?

* String interpolation in ml pipeline example has been corrected as per scala standard.

## How was this patch tested?
* manually tested.

### CODE REVIEW

**Comment by srowen** [comment]
OK, we can do this for consistency. In general we don't spend a ton of time on changes like this, because there is already way too much else to review. Please make one change only (this one) that updates all examples in one go.

**Comment by chetkhatri** [comment]
@srowen Absolutely correct, this all in one shot. I did changes in all.

**Comment by chetkhatri** [comment]
@srowen also i did merge another similiar PR with graphx to here. so Just  FYI  - we are good.

**Comment by srowen** [comment]
I think there are many more examples of this; a quick search suggests there are about 40 example files that have some string concatenation.

**Comment by chetkhatri** [comment]
In scala ? I don't think so. I am re-iterating and doing double check.

**Comment by chetkhatri** [comment]
You're correct - I missed other packages. I will re-confirm soon. Thanks.

**Comment by chetkhatri** [comment]
@srowen I rechecked all scala examples and this is commulative PR for the same.

File: examples/src/main/scala/org/apache/spark/examples/graphx/Analytics.scala [hunk 1]
```diff
@@ -145,9 +145,9 @@ object Analytics extends Logging {
           // TriangleCount requires the graph to be partitioned
           .partitionBy(partitionStrategy.getOrElse(RandomVertexCut)).cache()
         val triangles = TriangleCount.run(graph)
-        println("Triangles: " + triangles.vertices.map {
+        println(s"Triangles: ${triangles.vertices.map {
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/graphx/Analytics.scala]: This is probably more readable if the expression is stored in a local val first. Anything nontrivial like this gets hard to parse in an interpolated string. While we're here, worth fixing up.

File: examples/src/main/scala/org/apache/spark/examples/ml/ChiSquareTestExample.scala [hunk 2]
```diff
@@ -52,9 +52,9 @@ object ChiSquareTestExample {
 
     val df = data.toDF("label", "features")
     val chi = ChiSquareTest.test(df, "features", "label").head
-    println("pValues = " + chi.getAs[Vector](0))
-    println("degreesOfFreedom = " + chi.getSeq[Int](1).mkString("[", ",", "]"))
-    println("statistics = " + chi.getAs[Vector](2))
+    println(s"pValues = ${chi.getAs[Vector](0)}")
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/ChiSquareTestExample.scala]: I think this is OK; anything more complex I might suggest breaking out the expression into a val.

File: examples/src/main/scala/org/apache/spark/examples/ml/CorrelationExample.scala [hunk 3]
```diff
@@ -51,10 +51,10 @@ object CorrelationExample {
 
     val df = data.map(Tuple1.apply).toDF("features")
     val Row(coeff1: Matrix) = Correlation.corr(df, "features").head
-    println("Pearson correlation matrix:\n" + coeff1.toString)
+    println(s"Pearson correlation matrix:\n ${coeff1.toString}")
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/CorrelationExample.scala]: Another thing we could improve: `.toString` is redundant here I believe

File: examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala [hunk 4]
```diff
@@ -169,10 +169,10 @@ private class MyLogisticRegressionModel(
     Vectors.dense(-margin, margin)
   }
 
-  /** Number of classes the label can take. 2 indicates binary classification. */
+  // Number of classes the label can take. 2 indicates binary classification.
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala]: Yeah, good to make this a standard comment, not scaladoc style

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala [hunk 5]
```diff
@@ -31,12 +31,11 @@ object QuantileDiscretizerExample {
 
     // $example on$
     val data = Array((0, 18.0), (1, 19.0), (2, 8.0), (3, 5.0), (4, 2.2))
-    val df = spark.createDataFrame(data).toDF("id", "hour")
+    val df = spark.createDataFrame(data).toDF("id", "hour").repartition(1)
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: Although it looks weird, I think the author intended the repartition(1) to not appear in the body of the example that's copied into the docs. I wouldn't change this.

File: examples/src/main/scala/org/apache/spark/examples/streaming/CustomReceiver.scala [hunk 6]
```diff
@@ -82,9 +82,9 @@ class CustomReceiver(host: String, port: Int)
    var socket: Socket = null
    var userInput: String = null
    try {
-     logInfo("Connecting to " + host + ":" + port)
+     logInfo(s"Connecting to $host $port")
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/streaming/CustomReceiver.scala]: Nit: we could make the string consistent with the one two lines below by adding a colon

**Review by srowen — CHANGES_REQUESTED** [review]
A few small comments here but looking fine as an improvement. Also have a look for more old `.format()` calls if you like; we could update those.

File: examples/src/main/scala/org/apache/spark/examples/mllib/AssociationRulesExample.scala [hunk 7]
```diff
@@ -42,14 +42,13 @@ object AssociationRulesExample {
...
+        s" ${rule.confidence}")
     }
     // $example off$
 
     sc.stop()
   }
 
 }
-// scalastyle:on println
+// scalastyle:on println
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/mllib/AssociationRulesExample.scala]: It didn't like the lack of newline at the end of this file

File: examples/src/main/scala/org/apache/spark/examples/graphx/Analytics.scala (hunk 1 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/graphx/Analytics.scala]: sure

File: examples/src/main/scala/org/apache/spark/examples/ml/ChiSquareTestExample.scala (hunk 2 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/ChiSquareTestExample.scala]: Ok

File: examples/src/main/scala/org/apache/spark/examples/ml/CorrelationExample.scala (hunk 3 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/CorrelationExample.scala]: Addressed.

File: examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala (hunk 4 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/DeveloperApiExample.scala]: +1

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala (hunk 5 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: ok

File: examples/src/main/scala/org/apache/spark/examples/streaming/CustomReceiver.scala (hunk 6 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/streaming/CustomReceiver.scala]: Done, addressed

**Review by chetkhatri — COMMENTED** [review]
@srowen apologize for delay :) I have addressed suggestion. thanks

**Comment by chetkhatri** [comment]
@srowen please do re-run the build.

File: examples/src/main/scala/org/apache/spark/examples/ml/VectorIndexerExample.scala [hunk 8]
```diff
@@ -41,8 +41,8 @@ object VectorIndexerExample {
     val indexerModel = indexer.fit(data)
 
     val categoricalFeatures: Set[Int] = indexerModel.categoryMaps.keys.toSet
-    println(s"Chose ${categoricalFeatures.size} categorical features: " +
-      categoricalFeatures.mkString(", "))
+    println(s"Chose ${categoricalFeatures.size} " +
+      s"categorical features: {$categoricalFeatures.mkString(", ")}")
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/VectorIndexerExample.scala]: Typo here. Please make sure build/test/style passes locally before pushing again, or else this takes a lot of work to review.

Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/VectorIndexerExample.scala]: I did fixed this. Can you please give me steps as a check list before commit for test.

Comment (srowen) [reply examples/src/main/scala/org/apache/spark/examples/ml/VectorIndexerExample.scala]: `./dev/run-tests`

**Comment by chetkhatri** [comment]
@srowen Okey. current status looks good

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala [hunk 9]
```diff
@@ -45,7 +45,7 @@ object QuantileDiscretizerExample {
       .setNumBuckets(3)
 
     val result = discretizer.fit(df).transform(df)
-    result.show()
+    result.show(false)
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: One more question - is it necessary to make this not truncate?

File: examples/src/main/scala/org/apache/spark/examples/mllib/HypothesisTestingExample.scala [hunk 10]
```diff
@@ -68,7 +68,7 @@ object HypothesisTestingExample {
     // against the label.
     val featureTestResults: Array[ChiSqTestResult] = Statistics.chiSqTest(obs)
     featureTestResults.zipWithIndex.foreach { case (k, v) =>
-      println("Column " + (v + 1).toString + ":")
+      println(s"Column ${(v + 1).toString} :")
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/mllib/HypothesisTestingExample.scala]: .toString is redundant here and elsewhere with interpolation. I think that should be simplified.

File: examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala [hunk 11]
```diff
@@ -42,11 +42,11 @@ object LatentDirichletAllocationExample {
...
 
     // Output topics. Each is a distribution over words (matching word count vectors)
-    println("Learned topics (as distributions over vocab of " + ldaModel.vocabSize + " words):")
+    println(s"Learned topics (as distributions over vocab of ${ldaModel.vocabSize} words):")
     val topics = ldaModel.topicsMatrix
     for (topic <- Range(0, 3)) {
-      print("Topic " + topic + ":")
-      for (word <- Range(0, ldaModel.vocabSize)) { print(" " + topics(word, topic)); }
+      print(s"Topic $topic :")
+      for (word <- Range(0, ldaModel.vocabSize)) { print(s" ${topics(word, topic)}") }
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala]: Go ahead and put the print on a new line (I know it wasn't before)

**Review by srowen — CHANGES_REQUESTED** [review]

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala (hunk 9 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: We're following same style in other examples so it is good to do.

File: examples/src/main/scala/org/apache/spark/examples/mllib/HypothesisTestingExample.scala (hunk 10 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/mllib/HypothesisTestingExample.scala]: @srowen Thanks , Changes addressed

File: examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala (hunk 11 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala]: @srowen Thanks for suggestion, it has been addressed.

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala (hunk 9 above)
Comment (srowen) [reply examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: Which other examples? most do not set this, and the Java equivalent doesn't either. If there's a good reason that the output needs to be untruncated, that's fine, just also change the Java example.

File: examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala [hunk 12]
```diff
@@ -46,7 +46,10 @@ object LatentDirichletAllocationExample {
     val topics = ldaModel.topicsMatrix
     for (topic <- Range(0, 3)) {
       print(s"Topic $topic :")
-      for (word <- Range(0, ldaModel.vocabSize)) { print(s" ${topics(word, topic)}") }
+      for (word <- Range(0, ldaModel.vocabSize))
+          {
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala]: No, we put the open brace on the preceding line with for, and don't triple indent. See any other for loop in the code.

**Review by srowen — CHANGES_REQUESTED** [review]
Still more instances of .toString. Just search on the changes page, even, to see them.

File: examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala (hunk 12 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/mllib/LatentDirichletAllocationExample.scala]: @srowen sure done.

File: examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala (hunk 9 above)
Comment (chetkhatri) [reply examples/src/main/scala/org/apache/spark/examples/ml/QuantileDiscretizerExample.scala]: @srowen correct either way it works for ex. examples/ml/LDAExamples.scala

**Comment by chetkhatri** [comment]
@srowen Request for review when you get on this.

**Comment by srowen** [comment]
Merged to master
